### ISSUE TITLE
Add NotFound extension.

### ISSUE THREAD

**Issue by fiseni** [ardalis/GuardClauses#102]
It's a common scenario where we need to validate if a given queried object/entity exists. Instead of checking for null, we can have a more explicit extension for this scenario. We may name it simply `NotFound`.

The usage will be something like

```c#
int id = 1;
Customer myCustomer = customerRepository.GetByID(id);
Guard.Agsint.NotFound(id, myCustomer, nameof(myCustomer));
```

We have two options here:
- We can throw `KeyNotFoundException`
```c#
throw new KeyNotFoundException($"The querried object {objectName} with Key: {key} is not found!");
```
- We can add and expose a custom exception `NotFoundException` in the package.
```c#
public class NotFoundException : Exception
{
	public NotFoundException(string key, string objectName) 
		: base($"Queried object {objectName} was not found, Key: {key}")
	{
	}

	public NotFoundException(string key, string objectName, Exception innerException) 
		: base($"Queried object {objectName} was not found, Key: {key}", innerException)
	{
	}
}
```

**Comment by ardalis** [comment]
Which do you prefer?

**Comment by fiseni** [comment]
The custom exception, it's more explicit. 
You can then target the exception in try/catch and return [NotFound result](https://docs.microsoft.com/en-us/dotnet/api/microsoft.aspnetcore.mvc.controllerbase.notfound?view=aspnetcore-5.0) perhaps?

**Comment by ardalis** [comment]
Yes, or similar with Ardalis.Result if you don't want the exception to bubble up that far. Wondering if that's something I can build into that package, and how/if it should depend on this one for that exception type...

Want to PR this?

**Comment by fiseni** [comment]
If Ardalis.Result already depends on GuardClauses, then sure. I have to crosscheck what you have there and how this can be utilized.
But, I wouldn't define the same exception in both packages (in that case they are 2 different exceptions actually).

I'll add the PR later tonight or tomorrow if it's not an issue.
